pkg/infra/game/stage/trade/trade.go

```diff
@@ -0,0 +1,241 @@
+package trade
+
+import (
+	"fmt"
+	"infra/game/agent"
+	"infra/game/commons"
+	"infra/game/message"
+	"infra/game/stage/trade/internal"
+	"infra/game/state"
+	"infra/logging"
+	"time"
+
+	"github.com/benbjohnson/immutable"
+)
+
+// HandleTrade
+// A complete trading stage contains several rounds.
+// In each round, the following steps take place in order:
+// 1. Each agent can respond to one of the trading negotiations it is involved in OR propose a new trade to another agent.
+// 2. Main thread collects trade messages from all agents, and updated the state accordingly.
+// 3. Collected message will be forwarded to corresponding target agents in the start of next round.
+func HandleTrade(s state.State, agents map[commons.ID]agent.Agent, channelsMap map[commons.ID]chan message.TaggedMessage, round uint, roundLimit uint) {
+	// track offers made by each agent, no repeated offers are allowed
+	// i.e. only one offer of a specific item from an agent to another agent is allowed to exist simultaneously
+	availableWeapons := make(map[commons.ID][]state.Item)
+	availableShields := make(map[commons.ID][]state.Item)
+	// track all ongoing negotiations
+	negotiations := make(map[commons.TradeID]message.TradeNegotiation)
+	info := internal.NewInfo(negotiations, *internal.NewInventory(availableWeapons, availableShields))
+	// extract inventory from agents
+	for agentID, agentState := range s.AgentState {
+		info.Inventory.Weapons()[agentID] = commons.ImmutableListToSlice(agentState.Weapons)
+		info.Inventory.Shields()[agentID] = commons.ImmutableListToSlice(agentState.Shields)
+	}
+
+	for r := uint(0); r < round; r++ {
+		starts := make(map[commons.ID]chan interface{})
+		closures := make(map[commons.ID]chan interface{})
+		responses := make(map[commons.ID]chan message.TradeMessage)
+
+		for id, a := range agents {
+			a := a
+			agentState := s.AgentState[a.BaseAgent.ID()]
+			availWeapons := commons.SliceToImmutableList(info.Weapons()[id])
+			availShields := commons.SliceToImmutableList(info.Shields()[id])
+			requests := FindNegotiations(id, info.Negotiations())
+
+			start := make(chan interface{})
+			starts[id] = start
+			closure := make(chan interface{})
+			closures[id] = closure
+			response := make(chan message.TradeMessage)
+			responses[id] = response
+
+			go (&a).HandleTrade(agentState, availWeapons, availShields, &requests, start, closure, response)
```

Comment:
```suggestion
			start := make(chan interface{})
			starts[id] = start
			closure := make(chan interface{})
			closures[id] = closure
			response := make(chan message.TradeMessage)
			responses[id] = response

			go (&a).HandleTrade(
				s.AgentState[a.BaseAgent.ID()],
				commons.SliceToImmutableList(info.Weapons()[id]),
				commons.SliceToImmutableList(info.Shields()[id]), 
				&FindNegotiations(id, info.Negotiations()),
				start, closure, response)
```
We can probably inline the agentState,availWeapons, availShields and requests variable since they're only used in the coroutine.